Send email if reproducible built fails in the CI

ShahanaFarooqui opened the pull request:
Changelog-None.

ShahanaFarooqui commented:
Hi @s373nZ, Please review this PR to send an email notification if the CI fails during any reproducible build step.

I have also updated the folder location for this script from `/release` to `/repro` just to avoid confusion with the `release.yml`/`build-release.sh` script.

I am also considering merging `repro.yml` with `release.yml` in the future, because `repro.yml` serves as a pre-stage for `release.yml`. I would appreciate your thoughts on that too.

ShahanaFarooqui commented on .github/workflows/repro.yml:
```diff
@@ -31,6 +31,22 @@ jobs:
+          password: ${{ secrets.EMAIL_PASSWORD }}
+          subject: "Failure in Builder Image Setup Step for ${{ matrix.version }} - ${{ github.workflow }}"
+          to: sfarooqui@blockstream.com
+          from: sfarooqui@blockstream.com
```
Thanks, added `to` as variable `DISTRIBUTION_LIST`, `from` can be mapped with secret `EMAIL_USERNAME`.

ShahanaFarooqui commented:
@s373nZ 
> That said, your current approach looks pretty solid to me.

Yes, the goal is to send a customised email so that it stands out from the others and is not overlooked.

> I would suggest consolidating the three different action-send-email steps into one, by making it a separate job.

Thanks for pushing me to avoid being lazy 😄! I was not happy about repeating the step, but wanted to capture the details of the failed step as well. It took a little time, but the email and workflow are much cleaner now. I ended up merging them into a single step at the end.

> We could try reusing the repro.yml steps by utilizing a [reusable workflow or a composite action].

I would prefer to keep everything in one workflow. I plan to run the repro step on a scheduled basis, while the other steps (including the repro) will execute when a tag is pushed.

s373nZ commented:
@ShahanaFarooqui It just occurred to me that this line (in all the Ubuntu Dockerfiles) might cause a problem with changing the folder location to `/repro`:

https://github.com/ElementsProject/lightning/blob/2791c60bb22b8b0014b38942457025c892df04b7/contrib/reprobuild/Dockerfile.noble#L74

Since the files are reused in both the repro build process and the release process, it might cause an error.

ShahanaFarooqui commented:
>  It just occurred to me that this line (in all the Ubuntu Dockerfiles) might cause a problem with changing the folder location to /repro:

- Should this be an issue, considering that the `Dockerfile.noble` is used exclusively to build the `cl-repro-noble` image, and the next step only uses this newly created image? Isn't the `repro` folder mainly responsible for creating the `version.txt` and `git.log` files and changing user permissions?

OR

- Why did the action not fail from last night scheduler if this was an issue?

s373nZ commented:
> Should this be an issue, considering that the Dockerfile.noble is used exclusively to build the cl-repro-noble image, and the next step only uses this newly created image? Isn't the repro folder mainly responsible for creating the version.txt and git.log files and changing user permissions?

That `CMD` statement in the Dockerfile says that is the default command that is executed when you `docker run` the built image, so it is getting run to actually build the release here:

https://github.com/ElementsProject/lightning/blob/2791c60bb22b8b0014b38942457025c892df04b7/.github/workflows/repro.yml#L40

That line is copying the repro build archive to the `./release` directory, and the rest of the `repro.yml` steps look there to parse the filename here:

https://github.com/ElementsProject/lightning/blob/2791c60bb22b8b0014b38942457025c892df04b7/.github/workflows/repro.yml#L62

Also, IIRC I needed to create the `./release` directory in the CI because it doesn't exist after a fresh checkout.

One initial idea to get around this might be to try adding an `ARG` to the Dockerfile which defaults to `release` but pass in `repro` in this case. Probably the thing to do is run the action in a test branch set to trigger `on.branches.<test-branch-name>` just to be sure.

Hope this makes sense. LMK if it doesn't, or I'm missing something. I can help or chime back in tomorrow.

ShahanaFarooqui commented:
The action is still [successfully](https://github.com/ShahanaFarooqui/lightning/actions/runs/12149618856/job/33880686294) completing with the `/repro` directory. However, I have reverted the change and switched back to using the `/release` folder for now. This will not be a concern once we merge both workflows anyways.

s373nZ commented:
@ShahanaFarooqui I'm curious how it completed successfully w/o the `release` directory, but I can't see the build output from the run in the CI logs anymore. By grouping the commands and redirecting the output to log files, we would be sacrificing centralized observability in the Github Actions CI interface (for both successes and failures) in order to gain more context in the error email.

IMHO, I think the appropriate solution is to leave task commands as they were previously so we still get the log output in the UI, and just have a simpler email with less context that reports there was an error and provides a link to the CI output. The recipients can view the log output from Github actions in the case of an error, and others who are not on the `DISTRIBUTION_LIST` can also see the workflow output in cases of both success or failure. What do you think?

s373nZ commented on .github/workflows/repro.yml:
```diff
@@ -65,7 +101,7 @@ jobs:
 
           if [ -n "$(cat release/version.txt | sed -n '/-modded/p')" ] || \
-             [ -n "$(echo $releasefile | sed -n '/-modded/p')" ]
+            [ -n "$(echo $releasefile | sed -n '/-modded/p')" ]
```
nit: re-align the brackets vertically?

ShahanaFarooqui commented:
> IMHO, I think the appropriate solution is to leave task commands as they were previously so we still get the log output in the UI, and just have a simpler email with less context that reports there was an error and provides a link to the CI output. The recipients can view the log output from Github actions in the case of an error, and others who are not on the `DISTRIBUTION_LIST` can also see the workflow output in cases of both success or failure. What do you think?

Agree, I removed the error capturing code as it was causing more complexity than it was worth. I also tried using the `tee` command to write to both standard output and the file, but error handling at each step still added unnecessary complexity. It is better to rely on the receiver to check the details directly in the action itself.

Posting the `tee` error handling for future reference though:
```
sudo tar -C ${{ matrix.version }} -c . | docker import - ${{ matrix.version }} 2>&1 | tee command.log || exit_code=$?
if [ -n "$exit_code" ]; then
  echo "ERROR<<EOF" >> "$GITHUB_ENV"
  echo "$(cat command.log)" >> "$GITHUB_ENV"
  echo "EOF" >> "$GITHUB_ENV"
  exit 1
fi
```
```
Logs: <pre>${{ env.ERROR }}</pre><br/>
```

s373nZ reviewed (approved):
Very nice! LGTM :)

ACK https://github.com/ElementsProject/lightning/commit/f28587b78ba2d150a3725b15a51d273615ac8789

s373nZ commented:
One final observation - it looks like if all three repro builds failed, then three separate emails would be sent because the failure email step is a part of the matrix job, right? You could try to consolidate it down into one email by making that step a separate `job` as per https://github.com/ElementsProject/lightning/pull/7897#pullrequestreview-2475563876 (using `needs: ubuntu`) but it might not be straightforward to add the failing step name directly into the email body.

If maybe having multiple emails per run failure is fine for you, the current code still LGTM :+1:

ShahanaFarooqui commented:
>  You could try to consolidate it down into one email by making that step a separate job

For now, I would prefer to keep them separate since I am the only one on the distribution list :D. I will consider merging them if their frequency seems unnecessarily high. 

> but it might not be straightforward to add the failing step name directly into the email body

Capturing the failing step name should not be a big issue, as we can set it as output instead.
